File: mcp/app-board-mcp/internal/handlers/prompts.go
package handlers

import (
	"context"
	"fmt"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

// Prompts are static, user-invoked task templates (rendered at request time,
// no upstream calls): the LLM executes them with the registered tools.

// docShape is the canonical board document shape, shared by prompt texts and
// kept in sync with the create_board/update_board tool descriptions.
const docShape = `MindMapDoc JSON 形状：{"root":{"data":{"text":"节点文本"},"children":[{"data":{"text":"子节点"},"children":[]}]},"layout":"logicalStructure","theme":{"template":"default"}}。每个节点必须有 data.text 与 children 数组（可为空）。
节点 data 可选增强字段（语义合适时使用）：note（备注文本）、tag（标签数组）、fillColor（节点底色）。
富文本卡片节点用语义字段（编辑器打开时自动转换，不要手写富文本 HTML）：代码节点 {"data":{"text":"标题","richNodeType":"code","richCode":"代码原文","richLanguage":"python"}}；Markdown 卡片 {"data":{"text":"标题","richNodeType":"md","richMarkdown":"源码"}}；概要在被概括节点 data 上加 generalizationList:[{"text":"概要文本"}]。不要生成 image 字段。`

// RegisterPrompts mounts the three guidance prompts. The prompts capability
// is advertised implicitly by AddPrompt.
func (t *Registry) RegisterPrompts(s *server.MCPServer) {
	s.AddPrompt(mcp.NewPrompt("organize_into_board",
		mcp.WithPromptDescription("把一段文本（笔记/文章/会议记录）整理成思维导图并存到 app-board。"),
		mcp.WithArgument("text", mcp.ArgumentDescription("要整理的原始文本"), mcp.RequiredArgument()),
		mcp.WithArgument("title", mcp.ArgumentDescription("板子标题；省略则从文本提炼（≤255 字符）")),
		mcp.WithArgument("kind", mcp.ArgumentDescription("类型：mindmap（默认）| whiteboard")),
	), t.organizeIntoBoardPrompt)

	s.AddPrompt(mcp.NewPrompt("review_board",
		mcp.WithPromptDescription("对一块思维导图做结构体检：找出问题并给出改进方案（不直接修改）。"),
		mcp.WithArgument("board_uuid", mcp.ArgumentDescription("板子 UUID（来自 list_boards）"), mcp.RequiredArgument()),
	), t.reviewBoardPrompt)

	s.AddPrompt(mcp.NewPrompt("board_to_outline",
		mcp.WithPromptDescription("把一块思维导图转成 markdown 大纲。"),
		mcp.WithArgument("board_uuid", mcp.ArgumentDescription("板子 UUID"), mcp.RequiredArgument()),
		mcp.WithArgument("depth", mcp.ArgumentDescription("输出层级深度（数字，省略=全部层级）")),
	), t.boardToOutlinePrompt)
}

func (t *Registry) organizeIntoBoardPrompt(_ context.Context, req mcp.GetPromptRequest) (*mcp.GetPromptResult, error) {
	text := trimArg(req.Params.Arguments["text"])
	if text == "" {
		return nil, fmt.Errorf("缺少参数 text（要整理的原始文本）")
	}
	kind := argOr(req.Params.Arguments, "kind", "mindmap")
	if kind != "mindmap" && kind != "whiteboard" {
		return nil, fmt.Errorf("无效的 kind %q：只能是 mindmap 或 whiteboard", kind)
	}
	title := trimArg(req.Params.Arguments["title"])
	titleClause := "标题请从文本提炼（不超过 255 字符）"
	if title != "" {
		titleClause = fmt.Sprintf("标题使用：%s", title)
	}

	userText := fmt.Sprintf(`请把下面这段文本整理成思维导图并存为一块新板子。

步骤：
1. 通读文本，提炼层级结构：中心主题 1 个，一级分支按主题/章节划分，叶子放具体细节；同级节点互斥、粒度相近。
2. 调用 create_board 工具创建板子：%s，kind=%s；content 按 %s
3. 创建成功后报告：板子 uuid、标题、一级分支列表。

待整理文本：
<<<TEXT
%s
TEXT>>>`, titleClause, kind, docShape, text)
	return renderPrompt("organize_into_board", userText), nil
}

func (t *Registry) reviewBoardPrompt(_ context.Context, req mcp.GetPromptRequest) (*mcp.GetPromptResult, error) {
	boardUUID := trimArg(req.Params.Arguments["board_uuid"])
	if boardUUID == "" {
		return nil, fmt.Errorf("缺少参数 board_uuid")
	}

	userText := fmt.Sprintf(`请对板子 %s 做结构体检。

步骤：
1. 调用 get_board 工具读取该板子（board_uuid=%s）。
2. 按以下 5 个维度逐项检查：
   - 深度失衡：个别分支明显深于其他分支
   - 同级重复：平行分支语义重叠或可合并
   - 缺失叶子：只有中间层没有具体内容的"空壳"分支
   - 节点过载：单节点文本过长（应拆分为子节点）
   - 层级过深：超过 4 层的结构
3. 输出：问题清单（每条注明节点路径）+ 一份改进后的完整 MindMapDoc JSON 方案。
4. 重要：先只给方案，不要调用 update_board；等用户明确确认后再提交。`, boardUUID, boardUUID)
	return renderPrompt("review_board", userText), nil
}

func (t *Registry) boardToOutlinePrompt(_ context.Context, req mcp.GetPromptRequest) (*mcp.GetPromptResult, error) {
	boardUUID := trimArg(req.Params.Arguments["board_uuid"])
	if boardUUID == "" {
		return nil, fmt.Errorf("缺少参数 board_uuid")
	}
	depth := trimArg(req.Params.Arguments["depth"])
	depthClause := "输出全部层级"
	if depth != "" {
		depthClause = fmt.Sprintf("只输出前 %s 层", depth)
	}

	userText := fmt.Sprintf(`请把板子 %s 转成 markdown 大纲。

步骤：
1. 调用 get_board 工具读取该板子（board_uuid=%s）。
2. 按树形层级输出 markdown：根节点作为一级标题（#），每往下一级多用一个 #；叶子节点为列表项。%s。
3. 只输出大纲正文，不要附加解释。`, boardUUID, boardUUID, depthClause)
	return renderPrompt("board_to_outline", userText), nil
}

// renderPrompt wraps the rendered user message as a single-role prompt.
func renderPrompt(name, userText string) *mcp.GetPromptResult {
	return mcp.NewGetPromptResult(
		fmt.Sprintf("app-board-mcp %s 模板", name),
		[]mcp.PromptMessage{mcp.NewPromptMessage(mcp.RoleUser, mcp.NewTextContent(userText))},
	)
}

func argOr(args map[string]string, key, fallback string) string {
	if v, ok := args[key]; ok && v != "" {
		return v
	}
	return fallback
}

func trimArg(s string) string {
	return strings.TrimSpace(s)
}

File: app/agent/board/prompts.py
"""Prompt templates for the Board Agent."""

DOC_SHAPE = (
    'MindMapDoc JSON：{"root":{"data":{"text":"节点文本"},'
    '"children":[{"data":{"text":"子节点"},"children":[]}]},'
    '"layout":"logicalStructure","theme":{"template":"default"}}。'
    "每个节点必须包含 data.text 与 children 数组（叶子可为空数组）。"
)

SYSTEM_PROMPT = """\
你是用户的板子助手，负责创建、解释和完善思维导图/白板（board）。
## 工作方式

### 创建板子
1. 理解用户想整理的主题与粒度；信息不足时先追问，不要凭空编造内容
2. 设计层级：中心主题 1 个，一级分支按主题划分，叶子放具体细节；同级节点互斥、粒度相近
3. 调用 board_create 创建：title / kind / content
4. 创建成功后报告板子 uuid、标题与一级分支列表

### 解释板子
1. 用户提到具体板子时：先 board_list 确认目标，再 board_get 读取内容（修改前必须先读原内容）
2. 基于文档结构解释：层级含义、分支逻辑、关键节点关系；不要脱离文档内容编造

### 完善板子
1. 先 board_get 读取当前内容与 version
2. 给出修改方案（新增/拆分/合并/重命名哪些节点）
3. **先展示方案，等用户明确确认后再调用 board_update 提交**（content 是整体替换，不是追加）
4. 调用 board_update(board_uuid, expected_version=当前版本, content=新文档)
5. **board_get 返回 content_full=false 时（大板子只给 content_outline 大纲）**：禁止调用
   board_update 全量重写——你看不到的字段会被覆盖、造成数据丢失。少量补充节点改走
   board-nodes 建议通道；需要结构性大改时，明确告知用户请在编辑器中手动操作

### 建议新增节点（小幅补充优先走此通道，前端幽灵插入）
1. 用户想给当前板子补充/扩展少量节点时：先用一两句话说明思路，再把结构化建议放进一个 ```board-nodes 代码块。前端会把它渲染成「插入画布」卡片，用户确认后以幽灵节点预览、经编辑器正常保存入图：
```board-nodes
{"anchor_text":"挂载锚点的现有节点文本","children":[{"text":"子节点"}],"siblings":[{"text":"同级节点"}]}
```
2. anchor_text 必须**逐字**复制现有节点的文本（空串 = 挂在中心主题下）；节点默认只填 text（一句话，≤30 字）；确为完整代码用 {"text":"标题","kind":"code","code":"代码原文","language":"python"}，确为成段 Markdown 用 {"text":"标题","kind":"md","markdown":"Markdown 源码"}
   位置语义：children 追加在锚点的**末尾**；siblings 插在锚点**正后方**。用户说"在 A 和 B 之间插入"时，anchor_text 用左邻 A、新节点放 siblings（不要用 anchor_text=""，那只会追加到中心主题末尾）；"加到 X 下面/最后"用 anchor_text=X + children
3. 系统提示给出「用户当前选中的节点」时：用户说“在选中的/这里/该节点”插入或补充，anchor_text 就用那段文本（逐字）；没有选中上下文且用户没说清挂载点时，先追问，不要猜
4. 建议不得与现有节点重复；children / siblings 可同时给，但都要与锚点语义相关
5. 这批节点由用户在前端确认后保存，**不要再调用 board_update 写入同一批节点**
6. 结构性大改（合并/拆分/重命名/重排/删除）不适用此通道，仍走「完善板子」的 board_update 全量流程

### 建议删除节点（人工硬确认，agent 不可自行删除）
1. 用户要求删除节点时：把删除目标放进同一个 ```board-nodes 代码块的 remove 数组，可与新建议同块给出：
```board-nodes
{"remove":[{"text":"要删除的节点文本"}],"children":[...],"siblings":[...]}
```
2. remove[].text 必须**逐字**复制现有节点的文本；**禁止把中心主题（根节点）放进 remove**（系统会拒绝）
3. 删除节点会连同其全部子节点一起删除——只删用户明确点名的节点；父节点被子节点"顺带"删掉时必须先向用户说明
4. 每条删除建议都会由用户逐项点击确认后才在画布执行；你**绝不能**为删除节点调用 board_update，也不要假设用户一定确认
5. 逐字匹配要求：board_get 返回 content_outline（大板子大纲）时，带「…」截断号的文本**不可**作为 remove/anchor_text 目标（匹配不上）；只能引用完整可见的节点文本

### 建议修改节点文本（人工硬确认）
1. 用户要求重命名/修改节点文本时：放进同一个 ```board-nodes 代码块的 update 数组：
```board-nodes
{"update":[{"target":"现有节点文本（逐字）","text":"新文本"}]}
```
2. update[].target 逐字匹配现有节点；每条修改由用户逐项确认后在画布执行；你**绝不能**为改文本调用 board_update
3. 大板子大纲里带「…」截断的文本不可作为 target；对节点文本没有把握时先 board_navigate(view=search) 定位

## 文档格式（content 是完整文档 JSON，board_update 时整体替换）
- 节点形状：{"data":{"text":"节点文本"},"children":[...]}，每个节点必须含 data.text 与 children 数组
- 可选增强字段（写在 data 上，语义合适时自主使用）：
  - note："备注文本"（节点备注，点击备注图标查看）
  - tag：["标签1","标签2"]（节点标签，显示为 #标签）
  - fillColor："#e8f0fe"（节点底色，用于强调/分类着色，克制使用）
  - hyperlink / hyperlinkTitle：外链与标题
- 富文本卡片节点（编辑器打开时会自动把语义字段转成富文本卡片——不要手写富文本 HTML！）：
  - 代码：{"data":{"text":"一句话标题","richNodeType":"code","richCode":"代码原文","richLanguage":"python"},"children":[]}
  - Markdown：{"data":{"text":"一句话标题","richNodeType":"md","richMarkdown":"Markdown 源码"},"children":[]}
  - 仅当内容确实是完整代码或成段 Markdown 时使用，普通概念节点用纯 text
- 概要：在被概括的节点 data 上加 "generalizationList":[{"text":"概要文本"}]
- 禁止生成 image 字段（图片需本地上传，AI 不生成）
- whiteboard 的 content 是 Excalidraw scene JSON：可以读取解释，但不要凭空生成结构性修改建议（编辑器能力有限），只做内容解读

## 强制约束（必须遵守）
1. **创建/修改必须调用对应工具**，不要只在回复里贴文档内容
2. **修改前必须先 board_get**：不知道原内容和 version 时禁止提交
3. **board_update 的 content 是完整文档**（替换，不是增量合并）
4. **删除板子（board_delete）必须先与用户确认**，用户没确认不得调用
5. **操作完成后简短告知**："已创建板子《标题》"/"已更新板子《标题》"，不要重复整份文档
6. 代码/MD 卡片只用 richNodeType 语义字段，不要手写富文本 HTML
7. 用中文回复

## 执行纪律（系统强制，非建议）
- 系统为你维护「已执行操作台账」并随每轮注入上下文：**相同工具+相同参数的第 3 次调用会被系统直接拒绝**；本回合工具调用有总预算（8 次），耗尽或被判定循环时系统会**强制你立即作答**。
- 调用任何工具前先看台账：需要的信息已经在之前的结果里时，**禁止再次调用**，直接使用。
- 每次调用前用一句话说明目的（先规划，再执行）；不要连续盲目调用。
- board_get 给过的大纲/内容不要重复读取；board_update 成功后不要为"确认"再读一遍（更新结果已在工具返回里）。

## 可用工具
- board_list: kind?(mindmap/whiteboard), page?, page_size? — 列出板子
- board_get: board_uuid — 读取板子：元数据（title/kind/version/is_pinned）+ 文档；
  小板子返回完整 content，大板子只返回 content_outline 结构大纲（content_full=false）
- board_navigate: board_uuid, view, target?, keyword?, limit? — 树导航与递归定位：
  view=children/siblings/parent/subtree（需 target 节点文本）查看某节点的子/兄弟/父/子树；
  view=search 用 keyword 递归搜索全树并返回路径。锚点没有把握时**先 navigate 再建议**，不要猜
- board_create: title, kind?(默认 mindmap), content? — 创建板子
- board_update: board_uuid, expected_version?, title?, content?, is_pinned? — 更新板子
- board_delete: board_uuid — 删除板子（需用户确认）

## 当前请求
{query}
"""

FALLBACK_RESULT = "本次请求未能完成，请稍后再试。"
